app/javascript/controllers/select_with_auto_complete_controller.js

```diff
@@ -0,0 +1,499 @@
+        break;
+
+      case "Down":
+      case "ArrowDown":
+        if (this.filteredOptions.length > 0) {
+          if (altKey && this.#isClosed()) {
+            this.#open();
+          } else {
+            if (this.#isClosed()) {
+              this.#open();
+            }
+            if (this.filteredOptions.length > 1) {
+              this.#setOption(this.#getNextOption(this.option));
+            } else {
+              this.#setOption(this.firstOption);
+            }
+          }
+        }
+        flag = true;
+        break;
+
+      case "Up":
+      case "ArrowUp":
+        if (this.#hasOptions()) {
+          if (this.#isOpen()) {
+            this.#setOption(this.#getPreviousOption(this.option));
+          } else {
+            this.#open();
+            if (!altKey) {
+              this.#setOption(this.lastOption);
+            }
+          }
+        }
+        flag = true;
+        break;
+
+      case "Esc":
+      case "Escape":
+        if (this.#isOpen()) {
+          this.#close();
+        } else {
+          this.#setValue();
+        }
+        flag = true;
+        break;
+
+      case "Tab":
+        this.#close();
+        if (this.option) {
+          this.#setValue(this.option);
+        }
+        break;
+
+      case "Home":
+        this.comboboxTarget.setSelectionRange(0, 0);
+        flag = true;
+        break;
+
+      case "End":
+        const length = this.comboboxTarget.value.length;
+        this.comboboxTarget.setSelectionRange(length, length);
+        flag = true;
+        break;
+
+      default:
+        break;
+    }
+
+    if (flag) {
+      event.stopPropagation();
+      event.preventDefault();
+    }
+  }
+
+  #isPrintableCharacter(str) {
+    return str.length === 1 && str.match(/\S| /);
+  }
+
+  #onComboboxKeyUp(event) {
+    let flag = false;
+    const char = event.key;
+
+    if (this.#isPrintableCharacter(char)) {
+      this.filter += char;
+    }
+
+    if (event.key === "Escape" || event.key === "Esc") {
+      return;
+    }
+
+    switch (event.key) {
+      case "Backspace":
+        this.filter = this.comboboxTarget.value;
+        this.debouncedFilterAndUpdate();
+        flag = true;
+        break;
+
+      case "Left":
+      case "ArrowLeft":
+      case "Right":
+      case "ArrowRight":
+      case "Home":
+      case "End":
+        this.#setOption(null);
+        flag = true;
+        break;
+
+      default:
+        if (this.#isPrintableCharacter(char)) {
+          this.debouncedFilterAndUpdate();
+          flag = true;
+        }
+
+        break;
+    }
+
+    if (flag) {
+      event.stopPropagation();
+      event.preventDefault();
+    }
+  }
+
+  #onComboboxClick() {
+    if (this.#isOpen()) {
+      this.#close();
+    } else {
+      this.#open();
+    }
+  }
+
+  #onComboboxFocus() {
+    this.filter = this.comboboxTarget.value;
+    this.#filterOptions();
+    this.#setOption(null);
+  }
+
+  #onComboboxBlur() {
+    this.#setOption(null);
```

Comment:
You should probably reset the filter here.  If a user types a partial filter, blurs, then refocuses, the old filter remains in memory.

```javascript
#onComboboxBlur() {
  this.#setOption(null);
  this.filter = this.comboboxTarget.value; // Sync filter with actual value
}
```